### DISCUSSION TITLE
Function to import modules

### DISCUSSION THREAD

**Discussion by AlexEhn** [tellurianinteractive/Tellurian.Trains.ModulesRegistryApp#29] [Q&A] [answered]
We have a separate module repository with a lot of information. 

Is there an import function/API that can be used to create modules?

**Comment by fjallemark** [comment, marked as answer]
I have investigated import of modules and concluded that this is a quite extensive task for several reasons:
* The flat data structure of Excel sheets is hard to match against a structured database.
* The data to import is incomplete.
* Matching some values to a valid property in the Module Registry is impossible
* You need to create records for the module owners in order to add modules, and this is complex.
* You cannot import directly into the database tables, it must pass though the data entry layer of the system to ensure that data is valid.

The only strategy that will work is to *first* validate all imported data, and if there are any errors, point them out to the person making the import. The errors must be corrected, and then a new validation is made, probably finding new errors. Then, if the import data finally is correct, the import must anyway assume a lot of default values that have to be checked and corrected in the Module Registry. My conclusion is that it is better to manually enter modules/stations for several reasons:
* Implementing an import function as described above is an extensive task. I know, because I have made an XPLN-validator.
* Manually entering data will probably result in higher data quality, because the additional data in the Module Registry will be entered by someone than knows the reality of each module.

So the content of it all is that I judge it a bigger task to implement an import, than manually enter the data. I rather spend my time entering modules/stations for you, than using time to write and test an import function.

**Reply by AlexEhn** [reply]
Thanks alot for investigating this Stefan!

All your arguments are very valid.
